src/policy/compressor/forwarding.rs

```diff
@@ -0,0 +1,165 @@
+use crate::policy::compressor::GC_MARK_BIT_MASK;
+use crate::util::constants::MIN_OBJECT_SIZE;
+use crate::util::heap::MonotonePageResource;
+use crate::util::metadata::side_metadata::spec_defs::{COMPRESSOR_MARK, COMPRESSOR_OFFSET_VECTOR};
+use crate::util::metadata::side_metadata::SideMetadataSpec;
+use crate::util::{Address, ObjectReference};
+use crate::vm::object_model::ObjectModel;
+use crate::vm::VMBinding;
+use atomic::Ordering;
+use std::marker::PhantomData;
+use std::sync::atomic::AtomicBool;
+
+/// A finite-state machine which processes the positions of mark bits,
+/// and accumulates the size of live data that it has seen.
+///
+/// The Compressor caches the state of the transducer at the start of
+/// each block by serialising the state using [`Transducer::encode`];
+/// the state can then be deserialised using [`Transducer::decode`].
+#[derive(Debug)]
+struct Transducer {
```

Comment:
It is a bit hard to understand how this type works. More comments would help. For example, explain what each fields mean. Also you could comment `step` to say that `address` is always either the start of an object, or the end of an object, etc.